### PR TITLE
RFC: Add instance type settings design proposal

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes # 

**Description**

- Adds Instance Type settings design doc and RFC.

**How was this change tested?**

*

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note
None
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jonathan-innis** [comment]
#2129, #2180, #1490, #2129, #1803, #1995, and #2161 addressed by this design. RFC from issue owners and others that are interested in the design.

File: designs/instance-type-settings.md
Comment (njtran) [inline designs/instance-type-settings.md:7]: nit: Can we reference the issue instead of the pull request here? 
```suggestion
* Define [custom device resources](https://github.com/aws/karpenter/issues/1900) on an instance type
```

**Review by njtran — DISMISSED** [review]
LGTM. Just one nit

File: designs/instance-type-settings.md
Comment (ellistarn) [inline designs/instance-type-settings.md:249]: IIUC, this is the only field of the instancetype CRD that is capable of influencing user data. I'm a bit wary of this coupling, how it might limit our design decisions in the future, as well as the security implications. Put differently, InstanceType is currently a readonly type and you're proposing to give it write privileges. It currently isn't capable of modifying the configuration of nodes, which is a special privilege of the Provisioner. I'm comfortable using it for binpacking, since this relationship exists today. 

Here's a less controversial alternative that functions identically w/ how `type InstanceType interface` works today, and doesn't sacrifice any use cases.

1. The InstanceType CRD can expose `pods` as a value in `Allocatable`
2. The `maxpods` value can be hardcoded in the provisioner's kubelet configuration
3. Our binpacker will default to what the `InstanceType` says (which may be overriden by the CRD), but if the provisioner sets a value, that's what we'll use.
4. If the customer wants to compute a dynamic (w.r.t. instancetype) max pods value, they can do so in user data, and they just need to make sure that this function agrees with the values they've overridden in the instancetype CRD. 
5. If you need different dynamic max pods values for the same instance type but different provisioners, stop and rethink your architecture.

File: designs/instance-type-settings.md
Comment (ellistarn) [inline designs/instance-type-settings.md:129]:
```suggestion
Because Karpenter is cloud-agnostic, we need to keep this in mind when adding changes to the API surface. For the initial iteration of the `InstanceType` CRD, we will assume that the instance type will be selected by the `.metadata.name` field which will map to the well-defined Kubernetes node field `node.kubernetes.io/instance-type`.
```

File: designs/instance-type-settings.md
Comment (ellistarn) [inline designs/instance-type-settings.md:254]: I'd love for us to do the work and figure out containerd overhead in 2022, and then provide customers with an accurate runtime class.

File: designs/instance-type-settings.md
Comment (ellistarn) [inline designs/instance-type-settings.md:296]: Now that I see this, my recommendation elsewhere to have the field be `InstanceType.Spec.Allocatable` is inaccurate. Is it more correct to make it `InstanceType.Spec.Capacity` and then we substract off these additional 3 values that are controlled at the provisioner level?

**Comment by ellistarn** [comment]
🚀 🎸 🎸 🚀

File: designs/instance-type-settings.md
Comment (jonathan-innis) [reply designs/instance-type-settings.md:249]: > as well as the security implications
I'm not quite sure I get the security implications that you are trying to call out here. Maybe you could describe it further. I see that this changes some read/write assumptions about the `InstanceType` and how it interacts with the current code logic but I'm not quite sure how this has security implications for the project.

> Here's a less controversial alternative that functions
I think the other option to call out is that we can surface `podsPerCore` in the `kubeletConfiguration` in the `Provisioner` which would give users another method to have this be a dynamic value without the overhead of having to write custom `userData`. If we are going to say that `pods` in the `.sepc.capacity` can't affect `--max-pods`, I would rather just punt having `pods` included entirely.

File: designs/instance-type-settings.md
Comment (jonathan-innis) [reply designs/instance-type-settings.md:296]: Yes, you are completely right. This should be `Capacity` not `Allocatable`

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:97]:
```suggestion
1. Configure custom device resources for `c4.large`
```
I think that's what you mean.

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:135]:
```suggestion
One of the primary questions presented is whether we should create a one-to-one mapping between settings and a given instance type or whether we should allow the ability to create more complex structures using set relationships and label selectors. Options are listed below, combined with the pros and cons for each:
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:139]:
```suggestion
1. Create an `InstanceType` CRD that specifies a one-to-one relationship between the settings and the instance type
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:146]:
```suggestion
        2. Someone may want to define settings over an instance family without having to individually configure the instances that exist within that family
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:147]:
```suggestion
2. Create an `InstanceTypeSetting` CRD that contains a list of instance types that allow a one-to-many relationship between settings and instance types. These instance types could be selected either with a `selector` or `regex`
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:150]:
```suggestion
        2. Lower maintenance burden from maintaining fewer `InstanceTypeSetting` CRDs
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:154]:
```suggestion
        3. If we were to deny overlapping `InstanceTypeSettings`, then the feature of specifying a group of instance types may not be used that often, as configuration between instance types may rarely be shared
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:157]:
```suggestion
**Recommendation:** Use `InstanceType` CRD for initial iteration for one-to-one mappings before considering introducing more complex constructs to the user. While specifying rulesets might reduce the overhead burden for customers to keep track of separate `InstanceTypeSettings`, ultimately without an API surface to view exactly what settings that ruleset will produce, it will ultimately be confusing for users.
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:209]: Instead of "I", perhaps use "you" or "someone" in this paragraph.

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:239]:
```suggestion
    1. `kubeletConfiguration` can live at the `Provisioner` level. If there is need to specify flat `kubeReserved` or `systemReserved` values based on the user’s knowledge of the OS/AMI, they are able to set this at the Provisioner level
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:245]:
```suggestion
    2. Other per-pod overhead that may be tied to things like image size can be tied to the pod `ephemeral-storage` requests which will help Karpenter ensure that it accurately provisions pods to nodes while also ensuring that there is parity between the scheduling decisions of Karpenter and the kube-scheduler.
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:254]:
```suggestion
1. There is still [users asks](https://github.com/aws/karpenter/issues/1803) for calculating the kubeReserved resources value based on the GKE calculation of the kubeReserved resources value, as opposed to the EKS default which is based on pod density. Should we allow some flag on the provisioner that can determine which way to base the calculation on (i.e. `kubeReservedMemoryStrategy`)?
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:289]:
```suggestion
**🔑 Recommendation:** Based on the additional overhead and muddiness of putting `blockDeviceMappings` in the `InstanceType` CR, we will punt on block device mappings for now. For now, we will consider extending ephemeral-storage through some dynamic mechanism out of scope, though the hope is that we will have some mechanism as a future feature that calculates the combined needed ephemeral-storage based off of the pod requests, kubeReserved and systemReserved.
```

File: designs/instance-type-settings.md
Comment (chrisnegus) [inline designs/instance-type-settings.md:316]:
```suggestion
        1. Pricing information is instance-type specific which means we are necessarily requiring a one-to-one mapping between some CRD and an instance type. This creates less flexibility if a user wants to describe common settings across an instance family or across an architecture type.
```

**Review by chrisnegus — COMMENTED** [review]
This reads really well. I just have few small suggestions.

File: designs/instance-type-settings.md
Comment (AndreasSko) [inline designs/instance-type-settings.md:157]: For our use case (similar to #1900 where nodes offer fuse devices via the [smarter device manager](https://gitlab.com/arm-research/smarter/smarter-device-manager)) the recommended approach of having a 1-1 mapping would create quite a significant overhead and burden: For every instance type and size, we would need to create an `InstanceType` resource and configure it. Even with templating this is quite a bit of effort and seems prone to errors. Also, we would need to keep the list of available instance types up to date, increasing the maintenance work. So going with the second option sounds like a more sustainable approach to me personally.

Comment (jonathan-innis) [reply designs/instance-type-settings.md:157]: How would you get around having this burden? My assumption is that each instance type wouldn't have a static number of resources that match the `smarter-devices/fuse` so you would need some dynamic way to configure the quantities for each of the different instance types, or at least a collection of instance types.

This becomes a problem when extending out to other resources. Even assuming that there is a consistent quantity of resources for the `smarter-devices/fuse` resource across _all_ instance types, that certainly wouldn't be true for other extended resources.

> Also, we would need to keep the list of available instance types up to date
This works as an override mechanism, so you would still have those instance types available if they didn't have a matching CR for them. The only difference there would be that the extended resource wouldn't be mapped onto it, which would still be something that would need to be addressed with the other "collection-based" approach.

Comment (jonathan-innis) [reply designs/instance-type-settings.md:157]: Also, the other thing that we were thinking with the approach proposed in this design, is we would offer some tooling/guidance around how to programmatically generate these values into CRs, so that the maintenance burden is reduced.

File: designs/instance-type-settings.md
Comment (jonathan-innis) [reply designs/instance-type-settings.md:245]: Good catch 👍

**Comment by jonathan-innis** [comment]
Migrating a thread from #2404:

> My assumption is that all instance types wouldn't have the same number of extended resources

Let me picture our two use cases:

For a large videoconferencing system (up to several hundreds of nodes per cluster) we leverage the ALSA /dev/snd device exposed by the kernel (module). Obviously the amount of these kernel devices available per host are not in relation to an instance size or type and probably not even on CPU architecture, but rather on linux kernel config. For this amount of instances Karpenter is the answer, overcomimg AWS provider limits of the availablity of a certain type/size, so the loosely tied flexibilty of types and sizes possible by Karpenter is what we have been looking for. Now having to constrain on few types and sizes by the current manual declaration cancels the whole benefits out.
We have a container build farm that [requires /dev/fuse device](https://www.redhat.com/sysadmin/podman-inside-kubernetes). The number of fuse devices provided by the kernel is also not tied to instance type or size, per default 1000 or more. Also here it is impossible for us to define all instance types times instance sizes to match all provider offers.
Granted there are other use cases that I might not be aware of, but I am struggling to picture a case where a one-to-one relation of this config would shine. In our case it would be a management nightmare so to speek. Maybe this context was missing to explain why my simple https://github.com/aws/karpenter/pull/2161 is sufficient for us. Thank you for considering!

**Comment by jonathan-innis** [comment]
Summarizing most feedback that we've heard. It seems like we might want to consider a mechanism for "selecting" sets of instance types so that we can ease the customer burden of having to maintain a 1:1 mapping.

An alternative solution here may be to keep the 1:1 mapping and provide some documentation around manifest generation for these `InstanceType` CRDs.

Though this might be more confusing to reason about; in general, it will cause less overhead for users.

**Comment by haarchri** [comment]
any chance to see this in one of the coming releases ?

**Comment by universam1** [comment]
@jonathan-innis We would really like to move out of the fork of #2161 for a significant number of clusters. We are forced to where otherwise it is be a blocking no-go for Karpenter.
In case the final solution is pending longer, would you please reconsider merging #2161 for the time being, as it does provide an MVP and it is tested well with hundreds of nodes to be working as expected?
Thank you

**Comment by jonathan-innis** [comment]
@universam1 We've talked a bit about this internally. I think we're willing to take a short-term experimental path that would unblock you in the main project. Ideally, the hope is that we can get a more robust solution that should solve for all the problems in the RFC going forward, but right now we haven't put this at super high priority.

Here's the thoughts:
1. The long-term solution is still under design and is deproiritized for the time-being but one idea that we have is to expose the InstanceTypes offered by a cloudprovider as CRs on the cluster. This would allow you to either SSA changes to the InstanceTypes or to create mutating webhooks that could receive the InstanceTypes and then edit them with anything that you need injected into them (like additional extended resources). Thoughts on this one?
2. As a short-term solution, we’d like to add a ConfigMap that exposes any extended resources (on top of the known ones) through a mapping from instance types to a v1.ResourceList like the example below. This would be an experimental feature and would be highly likely to change at some point in the future in favor of something more akin to 1 that would be a more robust long-term solution. Would you be interested in contributing to this short-term solution?

```yaml
"c5.large": "{"fuse": "1", "other-extended-resource": "4"}"
```

**Comment by ellistarn** [comment]
Perhaps we can model it based off of the instance type struct? ```{ "name": "c5.large", "capacity": {"fuse": "1", "other-extended-resource": "4"} }```

**Comment by universam1** [comment]
Thank you @jonathan-innis for considering this request and taking into account several options to solve it. Very much appreciated!

We are certainly happy to contribute to a solution - if you can provide some more details where you would see it fit into and what you'd specifically like to see, that would help getting us started. 
No issue with alpha grade solution that is likely going to change.

One requirement though from our side is to be agnostic to instance type and size. It is impossible for us to predefine all variants we like to support ( _type_ x _size_ ).
A simple globbing or regex support in `aws.instanceTypeOverrides` matcher doesn't sound hard for me to add and should meet our needs just fine. Thank you very much for consideration!

**Comment by jonathan-innis** [comment]
> It is impossible for us to predefine all variants we like to support

Can you explain why this is the case? If you want all types of the size `.xlarge`, you could just get the instance type names from an `aws ec2 describe-instance-types` call and then include these in the settings using some programatic means.

> A simple globbing or regex support

I'm not really a fan of opening up this settings schema to regex-based pattern matching. It doesn't seem particularly explicit in its design and implementation and seems overly-complicated given what we are trying to implement.

**Comment by universam1** [comment]
> include these in the settings using some programatic m

The instances we permit the devs. to request or allow Karpenter to choose, sum up at roughly 600 types and versions. Say anything >1 core, >3GB, incl. amd64 or arm64. Having to manage that, means updating a huge list and/or keeping hundreds of developers informed about what instances are whitelisted or why not.
Offloading this type decision is the selling point of Karpenter, and we love it for that. Feels like this undermines the whole concept.

> seems overly-complicated given what we are trying to implement

Could you explain please why this is complicated to implement? I believe few lines of code would suffice. In contrast coding an API or generator for Karpenter definition is a huge effort to implement.
It would mean that Karpenter now submits to an external API that it would have to comply with, which could easily become an unpleasant dependency for the project!?

**Comment by jonathan-innis** [comment]
I'm less a fan of the regex solution and might be swayed toward doing a requirements-based solution that allows you to create disparate objects that contain requirements (of which instance-type name can be one) and capacity details with a merge semantic. Would this suite the case you are describing and be more maintainable for you?

```json
[
   {
      "requirements":[
         {
            "key":"karpenter.k8s.aws/instance-size""operator":"In""values":[
               "large"
            ]
         }
      ],
      "capacity":{
         "fuse":"1",
         "other-extended-resource":"4"
      }
   }
]
```

In the ConfigMap, it would look like

```yaml
"aws.instanceTypeOverrides": "[{"requirements":[{"key":"karpenter.k8s.aws/instance-size","operator":"In","values":["large"]}],"capacity":{"fuse":"1","other-extended-resource":"4"}}]"
```

**Comment by universam1** [comment]
> might be swayed toward doing a requirements-based solution

That's a great one! I think the list of instance sizes is not that of an issue to manage. In fact I could imagine a hack to invert to a "NotIn: small" if we really want to permit all sizes but those irrelevant ones.

Thank you!
How do you like us to proceed?

**Comment by jonathan-innis** [comment]
>  I think the list of instance sizes is not that of an issue to manage

On this same train of thought, what if there was a `resources` field in the Provisioner API that allowed you to specify the amount of extended resource that a Node that a Provisioner launched would have (we would assume this amount for inflight scheduling until the Node resources actually registered), do you feel like this type of solution would also work well for you? I'm assuming if you have a set of requirements that are something like "only use instance types that aren't of the small variety" that this would map naturally to Provisioner requirements, where you could add these extended resources.

```yaml
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  resources:
    requests:
      fuse: "2" # We assume this much until the node registers
      some.example.com/extended-resource: "500Mi" # We assume this much until the node registers
```

**Comment by universam1** [comment]
> what if there was a resources field

I like this semantic actually most, a single authoritative place for Karpenter config! 👏🏻 

> we would assume this amount for inflight scheduling until the Node resources actually registered

AFAIR Karpenter does not actively assign pods to nodes any more like it used to, leaving that to kubescheduler? Great, we'll only have to ensure that the `smarter-device-manager` daemonset is up in time and has patched the custom resources on the nodes, _before_ Karpenter enters a new loop for unscheduled pods.... I guess we can tame such a flapping via `startupTaints` to have a defined cool down delay!? I think this should work fine!

> I'm assuming if you have a set of requirements that are something like "only use instance types that aren't of the small variety" that this would map naturally to Provisioner requirements

Indeed, main part of our specs contain
```yaml
spec:
  requirements:
    - key: karpenter.k8s.aws/instance-size
      operator: NotIn
      values:
        - nano
        - micro
        - small
```
Great idea!

**Comment by jonathan-innis** [comment]
> before Karpenter enters a new loop for unscheduled pods

We handle this the concept of "node initialization". We wait for the node to go ready and have resources registered before we consider new pods for scheduling to new nodes so you shouldn't have to handle this cool down period yourself. Once we are aware that a resource _should_ appear on the node, we wait until we see it before we consider it empty or not eligible to schedule pods.

**Comment by haarchri** [comment]
thanks for this RFC - i wonder what is the current status ?

**Comment by jonathan-innis** [comment]
@haarchri We're keeping this open to keep the conversation going, since this is the largest item that is currently tracking and creating pointers to all the open issues. There's been a lot of discussion among the maintainers and contributors around how we should attack a lot of these disparate problems, so the RFC is still unsettled and probably won't be until the project settles and merges a lot of the other current feature work:
1. https://github.com/aws/karpenter-core/pull/176
2. https://github.com/aws/karpenter-core/pull/221

**Comment by jonathan-innis** [comment]
As mentioned [here](https://github.com/aws/karpenter/pull/2390#issuecomment-1456581131), there is still a lot of uncertainty on the plan of attack with how to handle custom instance type settings. The issues mentioned in this initial RFC will be addressed at a later time, with a rewritten RFC. Closing this for now.

**Comment by universam1** [comment]
@ellistarn @jonathan-innis Trying to get a feeling for the time to expect for your long-term solution. We are wondering if we should maintain a fork with an intermediate solution for the time being or if there are other options possible?
As an idea, could we guard this  #2390 implementation behind an alpha flag to unblock those users willing to take the risk for the time being? Would that be a feasable approach?

**Pull request closed without merging**
